odoo/addons/base/models/res_lang.py

```diff
@@ -209,57 +245,81 @@ def install_lang(self):
-        return self.with_context(active_test=True).search([('url_code', '=', url_code)]).code or url_code
+    # ------------------------------------------------------------
+    # cached methods for **active** languages
+    # ------------------------------------------------------------
+    @property
+    def CACHED_FIELDS(self) -> list[str]:
+        """ Return the fields to cache for the active languages
+        Please promise all these fields don't depend on other models and context
+        and are not translated.
+        """
+        return ['code', 'url_code', 'name', 'active', 'flag_image_url', 'direction', 'id', 'grouping', 'decimal_point', 'thousands_sep']
+
+    def _get_data(self, **kwargs: Any) -> LangData:
+        """ Get the language data for the **first** given field value in kwargs
+        For example, get_data(code='en_US') will return the LangData
+        for the res.lang record whose 'code' field value is 'en_US'
+
+        :param dict kwargs: {field_name: field_value, ...}
+                field_name is the first key in kwargs and in self.CACHED_FIELDS
+                Try to reuse used 'field_name's: 'id', 'code', 'url_code'
+        :return: Valid LangData if (field_name, field_value) pair is for an
+                **active** language. Otherwise, Dummy LangData which will return
+                `False` for any field
+        :rtype: LangData
+        :raise: KeyError if field_name is not in self.CACHED_FIELDS
+        """
+        field_name, field_value = next(iter(kwargs.items()))
+        return self._get_active_by(field_name)[field_value]
 
-    def _lang_get(self, code):
+    def _lang_get(self, code: str):
         """ Return the language using this code if it is active """
-        return self.browse(self._lang_get_id(code))
+        return self.browse(self._get_data(code=code).id)
 
-    def _get_active(self, code):
-        """ Return the given language code if active, else return ``"en_US"``.
+    def _get_valid(self, code: str) -> str:
+        """ Return the given language code if active, else return ``False``.
         This method is convenient to retrieve a meaningful translated field
         value from the database.
         """
-        if not code or not self._lang_get_id(code):
-            return 'en_US'
-        return code
-
-    @tools.ormcache('self.code', 'monetary')
-    def _data_get(self, monetary=False):
-        thousands_sep = self.thousands_sep or ''
-        decimal_point = self.decimal_point
-        grouping = self.grouping
-        return grouping, thousands_sep, decimal_point
-
-    @tools.ormcache('self.id')
-    def _get_cached_values(self):
-        self.ensure_one()
-        return {
-            'id': self.id,
-            'code': self.code,
-            'url_code': self.url_code,
-            'name': self.name,
-        }
-
-    def _get_cached(self, field):
-        return self._get_cached_values()[field]
+        return self._get_data(code=code).code
```

Comment:
Doesn't this beat the purpose of the function, since `_get_data` only returns data from active records?
So you will end up with a key error in the `LangDataDict`

Reply:
see the comment of `_get_data`
```py
"""
return: Valid LangData if (field_name, field_value) pair is for an
                **active** language. Otherwise, Dummy LangData which will return
                `False` for any field
"""
```
It avoids developers to handle KeyError for an inactive/invalid language code

Reply:
Unless I missed something, this implementation leads to a `KeyError`, that's what I meant

Reply:
Sorry but I really do not understand how it can work since this function will raise a `KeyError` if it is not active.

Reply:
if the lang is inactive you `_get_data(code='xx_xx')` returns `LangData.dummy_lang_data` which will return `False` for all fields
`LangData.dummy_lang_data.any_field == False`

`LangData`'s `__getattr__` uses its `__getitem__` and we override the `__getitem__` to prevent `KeyError`
```
def __getitem__(self, field: str) -> Any:
         return super().get(field, False)
```

Reply:
Ok thanks, also just tried it :+1: